### PR TITLE
Feature/framework mapping update

### PR DESCRIPTION
**Pull request by AdityaKhatri**
## Changes

* Add new framework mapping flow

## This PR doesn't introduce any:

- [ ] temporary files, auto-generated files or secret keys
- [ ] build works
- [ ] eslint issues
- [ ] typescript issues
- [ ] codegen errors
- [ ] `console.log` meant for debugging
- [ ] typos
- [ ] unwanted comments
- [ ] conflict markers

## This PR contains valid:

- [ ] permission checks
- [ ] translations

### CODE REVIEW

File: app/views/AnalyticalFramework/AssistedTagging/FrameworkTagRow/AddButton/index.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+const predictionGroupKeySelector = (prediction: PredictionTag) => prediction.group ?? 'Misc';
+
+interface Props {
+    disabled?: boolean;
+    existingMappings: CategoricalMappingsItem[] | undefined;
+    predictionTags: PredictionTag[] | undefined;
+    onTagSelect: (tag: string | undefined) => void;
+}
+
+function AddButton(props: Props) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/FrameworkTagRow/AddButton/index.tsx]: We could have a more specific name

File: app/views/AnalyticalFramework/AssistedTagging/FrameworkTagRow/AddButton/index.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        existingMappings,
+        predictionTags,
+        onTagSelect,
+    } = props;
+
+    const [addInputShown, showAddInput, hideAddInput] = useBooleanState(false);
+
+    const handleTagSelect = useCallback((newTag: string | undefined) => {
+        hideAddInput();
+        onTagSelect(newTag);
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/FrameworkTagRow/AddButton/index.tsx]: Let's invert the order.

File: app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/Matrix2dTagInput/index.tsx [hunk 3]
```diff
@@ -228,37 +181,47 @@ function Matrix2dTagInput(props: Props) {
...
-    const handleSubRowClick = useCallback((subRowKey: string) => {
-        if (!selectedTag) {
+    const handleColumnAdd = useCallback((columnKey: string, tagKey: string) => {
+        const selectedMappingsIndex = mappings?.findIndex((mapping) => (
+            tagKey === mapping.tag
+            && mapping.association.type === 'COLUMN'
+            && mapping.association.columnKey === columnKey
+        ));
+
+        if (isDefined(selectedMappingsIndex) && selectedMappingsIndex !== -1) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/Matrix2dTagInput/index.tsx:201]: It is a bit weird to check if selectedMappingsIndex is defined or not.

File: app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/Matrix2dTagInput/index.tsx [hunk 4]
```diff
@@ -267,50 +230,60 @@ function Matrix2dTagInput(props: Props) {
...
-    const handleSubColumnClick = useCallback((subColumnKey: string) => {
-        if (!selectedTag) {
+    const handleSubRowAdd = useCallback((subRowKey: string, tagKey: string) => {
+        const selectedMappingsIndex = mappings?.findIndex((mapping) => (
+            tagKey === mapping.tag
+            && mapping.association.type === 'SUB_ROW'
+            && mapping.association.subRowKey === subRowKey
+        ));
+
+        if (isDefined(selectedMappingsIndex) && selectedMappingsIndex !== -1) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/Matrix2dTagInput/index.tsx:255]: It is a bit weird to check if selectedMappingsIndex is defined or not.

File: app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/Matrix2dTagInput/index.tsx [hunk 5]
```diff
@@ -320,116 +293,138 @@ function Matrix2dTagInput(props: Props) {
...
 
-            onMappingsChange(newMappings, widget.id);
+    const handleSubColumnAdd = useCallback((subColumnKey: string, tagKey: string) => {
+        const selectedMappingsIndex = mappings?.findIndex((mapping) => (
+            tagKey === mapping.tag
+            && mapping.association.type === 'SUB_COLUMN'
+            && mapping.association.subColumnKey === subColumnKey
+        ));
+
+        if (isDefined(selectedMappingsIndex) && selectedMappingsIndex !== -1) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/Matrix2dTagInput/index.tsx:320]: It is a bit weird to check if selectedMappingsIndex is defined or not.

File: app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/OptionTypeTagInput/index.tsx [hunk 6]
```diff
@@ -45,93 +44,83 @@ function OptionTypeTagInput(props: Props) {
...
+        widget.id,
+        mappings,
+    ]);
+    const handleCellAdd = useCallback((cellKey: string, tagKey: string) => {
+        const selectedMappingsIndex = mappings?.findIndex((mapping) => (
+            tagKey === mapping.tag
+            && mapping.association.optionKey === cellKey
+        ));
+
+        if (isDefined(selectedMappingsIndex) && selectedMappingsIndex !== -1) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/OptionTypeTagInput/index.tsx:80]: It is a bit weird to check if selectedMappingsIndex is defined or not.

File: app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/OrganigramTagInput/index.tsx [hunk 7]
```diff
@@ -42,93 +40,84 @@ function OrganigramTagInput(props: Props) {
...
+        mappings,
+    ]);
+
+    const handleCellAdd = useCallback((cellKey: string, tagKey: string) => {
+        const selectedMappingsIndex = mappings?.findIndex((mapping) => (
+            tagKey === mapping.tag
+            && mapping.association.optionKey === cellKey
+        ));
+
+        if (isDefined(selectedMappingsIndex) && selectedMappingsIndex !== -1) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/WidgetTagList/OrganigramTagInput/index.tsx:77]: It is a bit weird to check if selectedMappingsIndex is defined or not.

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by samshara — APPROVED** [review]

**Review by tnagorra — APPROVED** [review]

**Pull request merged**